README.md
# Kronos

If you use this software, please cite our work:

``` r
citation("kronos")
```

    ## 
    ## To cite kronos in publications use:
    ## 
    ##   Bastiaanssen TFS, Leigh SJ, Tofani GSS, Gheorghe CE, Clarke G, Cryan
    ##   JF (2023) Kronos: A computational tool to facilitate biological
    ##   rhythmicity analysis. bioRxiv. doi:
    ##   https://doi.org/10.1101/2023.04.21.537503
    ## 
    ## A BibTeX entry for LaTeX users is
    ## 
    ##   @Article{,
    ##     title = {Kronos: A computational tool to facilitate biological rhythmicity analysis},
    ##     author = {Thomaz F S Bastiaanssen and Sarah-Jane Leigh and Gabriel S S Tofani and Cassandra E Gheorghe and Gerard Clarke and John F Cryan},
    ##     journal = {bioRxiv},
    ##     year = {2023},
    ##     doi = {10.1101/2023.04.21.537503},
    ##   }

The following document is adapted from the supplementary materials for
this manuscript.

## 0. Introduction

Here, we will demonstrate how to use the Kronos package to assess
circadian rhythms in biological data collected over the day. For this
demonstration we have adapted some data from our laboratory, currently
undergoing peer review (reference to come).

We will demonstrate three common examples of experimental design
currently encountered in circadian rhythms analysis:

1.  Circadian rhythm analysis of a single variable over a 24-hour period

2.  Circadian rhythm analysis of a single variable over a 24-hour period
    with two or more treatment groups

3.  Circadian rhythm analysis of omics (higher-dimensional) data over a
    24-hour period with two or more treatment groups

All R code used to transform, reorganise and plot the data is also shown
below to provide a toolkit for aspiring and veteran bioinformaticians
alike. It should be noted that some variables in the demonstration data
sets provided here have been manipulated to better demonstrate the
functionality of this package.

*At the end of this document, we have included a few excursions into
more advanced subjects we find useful, but that did not necessarily fit
in with the mainline analysis.*

### Code chunk: Install Kronos and load our libraries

``` r
#Install kronos
#install.packages("kronos") 

#Load relevant packages
library(kronos)
library(tidyverse)
library(ggplot2)

#Load prepared data stored in Kronos library
data("kronos_demo")
```

### Code chunk: Load our data tables

Data should be prepared in “long form” for use in Kronos; that is with
values repeating in the “Timepoint” column, which defines when data was
collected during the period, here a 24-hour cycle.

Our package includes example datasets that we will use in this tutorial
that are pre-formatted. You can rearrange your data into long form using
pivot_longer() or gather() from the tidyverse.

Since we’re using prepared data, we already loaded it using
`data(kronos_demo)`. You can see how an example of how the data is
prepared below:

``` r
head(groupdata)
```

    ##   Animal_ID Timepoint Treatment Variable_1
    ## 1         6         5         A  1.2789856
    ## 2         7         5         A  1.0606877
    ## 3         8         5         A  1.0497167
    ## 4         9         5         A  1.0533610
    ## 5        10         5         A  1.4590203
    ## 6        12         5         A  0.8408964

Here we have prepared the omics data set with a separate metadata file
as is common when working with omics data sets. A metadata file can be
generated using select() in tidyr, or metadata and data can be combined
using inner_join if they contain an identical column (both name and
contents). The omics data set used here has been central-log transformed
to account for its compositional nature (see our guide
<https://arxiv.org/abs/2207.12475> for easy centred-log transformation).

## 1. Analysing Rhythmicity in a Single Group

We will start with the most simple example: analysing circadian
rhythmicity in a single experimental group for one outcome variable of
interest. For this we use the kronos function:

``` r
output <- kronos(formula = Variable_1 ~ time(Timepoint), 
                 data = onevariable, 
                 period = 24, 
                 verbose = T, 
                 pairwise = F)
```

    ## [1] "Using the following model: Variable_1 ~ Timepoint_cos + Timepoint_sin"
    ## [1] "Using the following model: Variable_1 ~ (Timepoint_cos + Timepoint_sin)"

Here we use the formula `Outcome Variable ~ time(Time Variable)`, which
is the most simple model used by the kronos function. We specify the
period as 24 (this can be adjusted as appropriate for the data
analysed). By selecting `verbose=T`, you will be able to see the models
run by the kronos function: this becomes increasingly useful when you
run more complex models. Finally we select `pairwise=F` here, as there
are no groups to compare for differences in rhythms.

The kronos function returns a kronosOut object, containing several
pieces of data that can be accessed using handy ‘getter’ functions,
which we will describe below:

1). The *getKronos_input()* function fetches the data that the model is
based on, as well as the calculated cosine and sine components.

``` r
head(getKronos_input(output))
```

    ##   Variable_1 Timepoint unique_group Timepoint_cos Timepoint_sin
    ## 1 -0.4239923         5         TRUE      0.258819     0.9659258
    ## 2 -1.0311723         5         TRUE      0.258819     0.9659258
    ## 3 -0.8739002         5         TRUE      0.258819     0.9659258
    ## 4 -0.5896825         5         TRUE      0.258819     0.9659258
    ## 5 -0.4174538         5         TRUE      0.258819     0.9659258
    ## 6 -0.4052512         5         TRUE      0.258819     0.9659258

2). The *getKronos_fit()* function fetches the key details for the
generated model that may be useful for prediction, modelling and other
statistical applications.

``` r
getKronos_fit(output)
```

    ## 
    ## Call:
    ## lm(formula = formula, data = data)
    ## 
    ## Coefficients:
    ##   (Intercept)  Timepoint_cos  Timepoint_sin  
    ##       -0.4063         0.3914        -0.6886

3). The *getKronos_trace()* function returns all the data required for
graphing the sinusoid curve, which can either be used in our specialized
ggplot2 functions, or can be used in other graphing packages. The
*y_hat* column represents the predicted value of the outcome variable:
this is essential for plotting the predicted sinusoid curve.

``` r
head(getKronos_trace(output))
```

    ##   Timepoint Timepoint_cos Timepoint_sin       y_hat unique_group
    ## 1      0.00     1.0000000    0.00000000 -0.01493121         TRUE
    ## 2      0.25     0.9978589    0.06540313 -0.06080489         TRUE
    ## 3      0.50     0.9914449    0.13052619 -0.10815799         TRUE
    ## 4      0.75     0.9807853    0.19509032 -0.15678776         TRUE
    ## 5      1.00     0.9659258    0.25881905 -0.20648595         TRUE
    ## 6      1.25     0.9469301    0.32143947 -0.25703975         TRUE

4). The *getKronos_groupwise()* function arguably fetches the most
useful output: this provides us with the p-value (p.val) and proportion
of the variance in the data explained (r.sq) when we fit our sinusoid
curve. Additionally we obtain the acrophase (acro) and amplitude of the
predicted curve, which can be used in our graphics functions to
visualise changes in curve with interventions (see `gg_kronos_circle`,
explored further below).

``` r
getKronos_groupwise(output)
```

    ##   unique_group        p.val      r.sq        avg     acro amplitude
    ## 1         TRUE 2.237134e-05 0.5345906 -0.4022588 19.97413  0.792033

### Figures

The package contains custom ggplot2 figure functions, that utilise the
kronos output to rapidly produce figures that convey important
information for circadian rhythms:

1.  `gg_kronos_circle()` generates a plot showing the acrophase and
    amplitude of the predicted curve, allowing the reader to rapidly
    access summary data regarding variables of interest, and to compare
    the summary data between groups in more complex models.At baseline,
    non-significant outcome measures are presented using dashed lines.
2.  `gg_kronos_sinusoid()` generates a x-y plot showing the outcome
    variable across the defined period. These graphs are useful for
    visualising the differences between specific timepoints assessed.

``` r
gg_kronos_circle(output)
```

![](README_files/figure-gfm/figures-1.png)<!-- -->

``` r
gg_kronos_sinusoid(output)
```

![](README_files/figure-gfm/figures-2.png)<!-- -->

## 2. Comparing Rhythmicity for More than Two Groups

Next we will demonstrate one of the unique features of the Kronos
package: the ability to compare circadian rhythms between more than two
groups. This is increasingly important as the use of complex
experimental designs grows in biological science. This example comprises
of three independent groups and is similar in setup to a one-way ANOVA.
For examples of more complex designs, see Excursion 1.

``` r
output2 <- kronos(formula = Variable_1 ~ Treatment + time(Timepoint), 
                  data = groupdata, 
                  period = 24, 
                  verbose = T, 
                  pairwise = T)
```

    ## [1] "Using the following model: Variable_1 ~ Treatment + Timepoint_cos + Timepoint_sin + Treatment:Timepoint_cos +     Treatment:Timepoint_sin - 1"
    ## [1] "Using the following model: Variable_1 ~ (Timepoint_cos + Timepoint_sin)"
    ## [1] "Using the following model: Variable_1 ~ (Timepoint_cos + Timepoint_sin)"
    ## [1] "Using the following model: Variable_1 ~ (Timepoint_cos + Timepoint_sin)"
    ## [1] "Fitting pairwise models"
    ## [1] "Using the following model: Variable_1 ~ unique_group * (Timepoint_cos + Timepoint_sin)"
    ## [1] "Using the following model: Variable_1 ~ unique_group * (Timepoint_cos + Timepoint_sin)"
    ## [1] "Using the following model: Variable_1 ~ unique_group * (Timepoint_cos + Timepoint_sin)"

``` r
gg_kronos_circle(output2)
```

![](README_files/figure-gfm/figures_complex-1.png)<!-- -->

``` r
gg_kronos_sinusoid(output2)
```

![](README_files/figure-gfm/figures_complex-2.png)<!-- -->

There are a few changes to the output generated by the kronos function:

1.  The *getKronos_groupwise()* output now contains a line for each of
    our groups.

``` r
getKronos_groupwise(output2)
```

    ##   unique_group       p.val       r.sq      avg      acro  amplitude
    ## 1            A 0.031490079 0.21886392 1.099699 12.681641 0.17966767
    ## 2            B 0.002912094 0.33147505 1.246347 18.592663 0.30283851
    ## 3            C 0.723230482 0.02287902 1.337834  3.447684 0.05150576

In this example you can see that groups A and B exhibit statistically
significant rhythms, while the model fitted to group C is
non-significant.

2.  We can now generate *pairwise models* output using `pairwise=T`.
    This generates pairwise comparisons between each of the groups:

``` r
getKronos_pairwise(output2)
```

    ## $`A vs B`
    ## Analysis of Variance Table
    ## 
    ## Response: Variable_1
    ##                            Df Sum Sq Mean Sq F value  Pr(>F)   
    ## unique_group                1 0.3626 0.36262  4.3279 0.04200 * 
    ## Timepoint_cos               1 0.1323 0.13228  1.5788 0.21406   
    ## Timepoint_sin               1 0.8859 0.88585 10.5728 0.00193 **
    ## unique_group:Timepoint_cos  1 0.4090 0.40901  4.8816 0.03118 * 
    ## unique_group:Timepoint_sin  1 0.5492 0.54916  6.5543 0.01314 * 
    ## Residuals                  57 4.7758 0.08379                   
    ## ---
    ## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
    ## 
    ## $`A vs C`
    ## Analysis of Variance Table
    ## 
    ## Response: Variable_1
    ##                            Df Sum Sq Mean Sq F value    Pr(>F)    
    ## unique_group                1 0.9067 0.90674 14.3827 0.0003677 ***
    ## Timepoint_cos               1 0.1669 0.16692  2.6476 0.1093177    
    ## Timepoint_sin               1 0.0007 0.00068  0.0108 0.9174232    
    ## unique_group:Timepoint_cos  1 0.3425 0.34246  5.4321 0.0233965 *  
    ## unique_group:Timepoint_sin  1 0.0390 0.03899  0.6185 0.4349195    
    ## Residuals                  56 3.5305 0.06304                      
    ## ---
    ## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
    ## 
    ## $`B vs C`
    ## Analysis of Variance Table
    ## 
    ## Response: Variable_1
    ##                            Df Sum Sq Mean Sq F value  Pr(>F)   
    ## unique_group                1 0.1279 0.12786  1.5595 0.21685   
    ## Timepoint_cos               1 0.0487 0.04868  0.5937 0.44417   
    ## Timepoint_sin               1 0.5623 0.56230  6.8581 0.01129 * 
    ## unique_group:Timepoint_cos  1 0.0025 0.00251  0.0306 0.86179   
    ## unique_group:Timepoint_sin  1 0.8940 0.89402 10.9038 0.00166 **
    ## Residuals                  57 4.6735 0.08199                   
    ## ---
    ## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Above we can see that overall group A is significantly different between
B and C, and that group B exhibits a significantly different rhythm from
A and C.

3.  When including independent variables, kronos will also calculate an
    overall interaction with time which can be accessed using the code
    below:

``` r
getKronos_pairwise_p(output2)
```

    ##             adj.p.val
    ## Treatment 0.004532107

This is calculated by performing a Bonferroni correction on the
interactions between both the sine and cosine time components and the
independent variable. The p-value reported is the lowest following
correction.

## 3. ’Omics Analysis

Now we will demonstrate how to adapt the kronos package for ’omics
analysis, where there are many outcome variables. We have written the
`fw_kronos` (feature-wise) function specifically for this purpose. This
function behaves very similar to the main kronos function.

It requires two core types of data: First, a table of data with rows as
features and columns as samples as input. Make sure that the feature
labels are row names rather than a column. Second, a metadata table with
rows as samples and metadata entries as columns. Suitable input data
looks like this:

``` r
#plot a little of how a big omics data should be formatted
head(bigdata, n = c(5, 5))
```

    ##                   X49        X51        X52        X24        X25
    ## Variable_1  0.6997887  0.7510796  0.4496661  0.7070641  0.7314069
    ## Variable_2  0.9185654  0.3832566  0.5772013  0.2469332  1.1264888
    ## Variable_3  5.7376737  5.7759893  5.8397621  4.6582202  4.6705684
    ## Variable_4 -0.3160217  0.2157887  5.9146992  0.1599933  0.2605589
    ## Variable_5 -4.8665730 -4.5445920 -3.5170020 -4.6925208 -4.3942593

``` r
#plot a little of the metadata should be formatted
head(bigmeta)
```

    ##   Animal_ID Group Timepoint
    ## 1       X49     B         5
    ## 2       X51     B         5
    ## 3       X52     B         5
    ## 4       X24     B         5
    ## 5       X25     B         5
    ## 6       X26     B         5

From the user’s perspective, `fw_kronos` reads very similar to the main
`kronos` function. The core difference is that `fw_kronos` accepts data
and metadata as separate objects. Furthermore, `fw_kronos` doesn’t need
a term on the left-hand side of the formula. It will automatically,
sequentially apply the given formula with each feature in the input
table as the response variable.

``` r
out_list = fw_kronos(x = bigdata, 
                     formula = ~ Group + time(Timepoint), 
                     metadata = bigmeta, 
                     period = 24,
                     verbose = F,
                     pairwise = T) 
```

Now we have a list of kronosOut objects, which contain all our results.
This can be cumbersome to do manually, so we wrote the
`kronosListToTable` function for this purpose:

``` r
fit_df = kronosListToTable(out_list)

write.csv(fit_df, "README_files/RhythmicityResults.csv")
```

This will generate a csv containing the individual rhythmicity
calculations for the whole data set with an FDR correction to account
for multiple tests.

The resulting csv can be found
[here](https://github.com/thomazbastiaanssen/kronos/blob/main/README_files/RhythmicityResults.csv).

``` r
#plot a little of fit_df
head(fit_df, n = c(5, 5))
```

    ##                 B_p.val     A_p.val   C_p.val      B_r.sq     A_r.sq
    ## Variable_1 9.532798e-01 0.030901063 0.7771550 0.003294345 0.21991674
    ## Variable_2 9.297825e-01 0.002684955 0.3122346 0.005008424 0.34483205
    ## Variable_3 2.784019e-07 0.755935357 0.4488660 0.646891288 0.01978728
    ## Variable_4 8.603783e-01 0.009861269 0.4818005 0.010317651 0.28103213
    ## Variable_5 7.528003e-01 0.012385930 0.9204081 0.019392619 0.26923016

We can use a similar approach to obtain other components of the
kronosOut objects. Below we include the code to obtain the pairwise
comparisons as a single csv, as this is slightly more difficult.

``` r
#Create an empty container list of the appropriate length
pairwise_list = vector(mode = "list", length = length(out_list)) 

#The for-loop below generates a list containing the pairwise test results
for(m in 1:length(out_list)){
  pairwise_list[[m]] <- out_list[[m]]@pairwise_models
  names(pairwise_list)[m] <- names(out_list)[m]
  
}

#Generate a single bound list
bound_list <- lapply(X = pairwise_list, FUN = function(x){do.call(rbind, x)})

#collapse the bound list for each variable into a single dataframe 
pairwise_df <- do.call(rbind, bound_list)

#separate the comparison and the effects to make results more readable
pairwise_csv = pairwise_df %>% 
  rownames_to_column("ID") %>%
  separate(col = ID, into = c("Feature","ID"), sep = "\\.", extra = "merge") %>%
  separate(col = ID, into = c("Comparison","Effect"), sep = "\\.")
  


NEWS.md
# kronos 1.0.0.9000

* Added a `NEWS.md` file to track changes to the package.


R/kronos.R
#' Fit a cosinor model and extract relevant parameters
#' @description This is the main workhorse function in the 'kronos' package. It manages the individual functionalities of 'kronos', including rhythmicity analysis and differential rhythmicity.
#' @param data input data
#' @param formula A formula. Use the \code{time} function to designate which variable represents time. 
#' @param time A string. Should be the column name containing the time values.  
#' @param period A numeric. The length of a period, in the same format as the \code{time} variable  
#' @param pairwise A boolean. Toggles whether to perform pairwise ANOVAs as a TukeyHSD-like post-hoc. 
#' @param verbose A boolean. Toggles whether to print diagnostic information while running. Useful for debugging errors on large data sets.
#' @return A \code{kronosOut} S4 object containing coefficients and all operations.
#' @examples
#' #Load prepared data stored in Kronos library
#' data("kronos_demo")
#' output <- kronos(formula = Variable_1 ~ time(Timepoint), 
#' data = onevariable, period = 24, verbose = TRUE, pairwise = FALSE)
#' 
#' #Extracting data from the output object:
#' getKronos_fit(output)
#' getKronos_trace(output)
#' getKronos_groupwise(output)
#' 
#' #Plotting:
#' gg_kronos_circle(output)
#' gg_kronos_sinusoid(output)
#' 
#' #For high-dimensional data, use fw_kronos:
#' out_list = fw_kronos(x = bigdata[1:50,], formula = ~ Group + time(Timepoint), 
#' metadata = bigmeta, period = 24, verbose = FALSE, pairwise = TRUE) 
#'
#' #Extracting data from the output object: 
#' kronosListToTable(out_list)
#' 
#' 
#' #Plotting:
#' gg_kronos_acrogram(out_list)
#' 
#' @importFrom methods new
#' @importFrom stats anova as.formula coef lm pf predict terms update.formula get_all_vars
#' @importFrom utils combn
#' @export
kronos <- function(formula, data, time = NULL, period = 24, verbose = TRUE, pairwise = TRUE){

if(!is.null(time)){
  formula = as.formula(paste0(Reduce(paste, deparse(formula)), " + time(", time, ")"))
}    
#Clean and standardize input data
data <- get_all_vars(formula = formula, data = data)
  
#Identify time component
var_out  <- get_vars(formula = formula, time = time, data = data, verbose = verbose)  
time     <- var_out$time
response <- var_out$response
factors  <- var_out$factors

#convert all non-time and non-response variables to categorical data. 
data[,{{factors}}] <- sapply(data[,{{factors}}], "as.character")

if(is.null(factors)){data$unique_group <- TRUE}

if(!is.null(factors) & length(factors) == 1){
  data$unique_group = data[,{{factors}}]
}

if(!is.null(factors) & length(factors) > 1){
data$unique_group = do.call(what = "paste", c(data[,{{factors}}], sep = "_"))
}
#Set up sine and cosine component
data <- cbind(data, get_cos_sine(data = data[,time], colnamePrefix = paste0(time, "_"), period = period))

#Update formula to reflect sine and cosine component
formula <- build_kronos_formula(formula = formula, time = time, verbose = verbose)

#Fit the general model
fit  <- fit_cosinor_model(formula = formula, data = data, time = time, verbose = verbose, for_pw = FALSE)

#Predict values for plotting
vals <- kronos_predict(fit = fit, period = period, time = time, factors = factors, verbose = verbose)

#In case of no non-time factors
if(length(fit$xlevels) == 0){
  fit$model$unique_group = TRUE
  groupwise = vector(mode = "list", length = 1)
  groupwise[[1]] <- fit_groupwise_model(data = fit$model, group = TRUE, 
                                             time = time, period = period, verbose = verbose)
}

#In case of multiple groups
if(length(fit$xlevels) == 1){
#fit$model$unique_group <- paste(fit$model[,names(fit$xlevels)])

fit$model$unique_group <- do.call(what = "paste", list(data[,{{factors}}], sep = "_"))[!is.na(data[,response])]

groupwise = vector(mode = "list", length = length(unique(fit$model$unique_group)))

for(g in 1:length(groupwise)){
  groupwise[[g]] = fit_groupwise_model(data = fit$model, 
                                       group = unique(fit$model$unique_group)[g], 
                                       time = time, period = period, verbose = verbose)
  }
}

#In case of multiple groups
if(length(fit$xlevels) > 1){
  #fit$model$unique_group <- paste(fit$model[,names(fit$xlevels)])
  
  fit$model$unique_group <- do.call(what = "paste", c(data[,{{factors}}], sep = "_"))[!is.na(data[,response])]
  
  groupwise = vector(mode = "list", length = length(unique(fit$model$unique_group)))
  
  for(g in 1:length(groupwise)){
    groupwise[[g]] = fit_groupwise_model(data = fit$model, 
                                         group = unique(fit$model$unique_group)[g], 
                                         time = time, period = period, verbose = verbose)
  }
}


bygroup = do.call(rbind, groupwise)
row.names(bygroup) <- NULL

pairwise_models <- list("pairwise was not activated.")
if(pairwise){  
  pairwise_models <- list()
  if(verbose){print("Fitting pairwise models")}
  pairwise_models = pairwise_cosinor_model(data = fit$model, formula = formula, time = time, verbose = verbose)
}

pairwise_p_vals <- "pairwise was not activated."
if(pairwise){  
  pairwise_p_vals <- kronos_anova(fit = fit, time = time)
}

#initialize output container object
output = new("kronosOut", 
             input           = data,
             fit             = fit,
             to_plot         = vals,
             ind_fit         = bygroup, 
             pairwise_models = pairwise_models, 
             pairwise_p_vals = pairwise_p_vals, 
             plot_info       = list(time = time, response = response, period = period))

return(output)
}

#' Figure out what variable represents time. Called by main 'kronos' function. 
#' @description Extracts time from the formula and from the time argument. Also handles inconsistencies. 
#' @param formula A formula. Use the \code{time} function to designate which variable represents time. 
#' @param time A string. Should be the column name containing the time values.  
#' @param data input data
#' @param verbose A boolean. Toggles whether to print diagnostic information while running. Useful for debugging errors on large data sets.
#' 
get_vars <- function(formula, time, data, verbose = verbose){
  #extract time from formula
  modelterms <- terms(formula, specials = "time", keep.order = TRUE)
  
  model_time <-  all.vars(formula)[unlist(attr(modelterms, "specials"))]
  
  #If both formula and argument specifiy time, check them out. 
  if(!is.null(time) & !identical(model_time, character(0))){
    stopifnot("The time argument was stated in the formula and differs from the one given as the time argument" = identical(time, model_time))
  }
  
  #Preferrably use time from the formula. 
  if(!identical(model_time, character(0))){
    time <- model_time
  }
  #Check whether time argument exists in dataset
  stopifnot("The time argument argument was not found. Use the `time = <here>` argument or assign within the formula: `y ~ x + time(<here>)" = length(time) == 1)
  stopifnot("The time argument does not match any column in the data set." = time %in% colnames(data))
  
  model_response <-  all.vars(formula)[unlist(attr(modelterms, "response"))]
  model_factors  <-  all.vars(formula)[-c(unlist(attr(modelterms, "response")), 
                                          unlist(attr(modelterms, "specials")))]
  if(identical(character(0), model_factors)){
    model_factors <- NULL
  }
  return(list(time = time, response = model_response, factors = model_factors))
}

#' Get sine and cosine components
#' @description Based on 'cosinor' and 'limorhyde' packages
#' @param data input data
#' @param period A numeric. The length of a period, in the same format as the \code{time} parameter.  
#' @param colnamePrefix A character string. Typically the name of the "Time" variable. 
#' 
get_cos_sine <- function(data, period, colnamePrefix = NULL){
  b = cbind(cos(data/period * 2 * pi), 
            sin(data/period * 2 * pi))
  colnames(b) = c("cos", "sin")
  colnames(b) = paste0(colnamePrefix, colnames(b))
  return(b)
}

#' Fit cosinor model
#' @description Fit cosinor model for totality of data
#' @param formula A formula. Use the \code{time} function to designate which variable represents time. 
#' @param data Input data
#' @param time A string. Should be the column name containing the time values.  
#' @param verbose A boolean. Toggles whether to print diagnostic information while running. Useful for debugging errors on large datasets.
#' @param for_pw A boolean. Toggles whether to perform pairwise ANOVAs as a TukeyHSD-like post-hoc. 
#' 
fit_cosinor_model <- function(formula, data, time = NULL, verbose = verbose, for_pw = FALSE){
  #fix intercept for pairwise fits
  
  if(!for_pw & length(all.vars(formula)) > 3){
  formula = update.formula(formula, formula(paste0("~ . -1")))
  }
  
  #in case of singular group
  if(verbose){print(paste0("Using the following model: ", Reduce(paste0, deparse(formula))))}
  
  fit = lm(formula, data = data)
  
  return(fit)
}

#' Give tracing information for plotting purposes
#' @description Generate data needed to plot cosinor trace line. 
#' @param fit A model fit
#' @param time A string. Should be the column name containing the time values.  
#' @param factors A vector. The names of the independent variables. 
#' @param period A numeric. The length of a period, in the same format as the \code{time} parameter.  
#' @param verbose A boolean. Toggles whether to print diagnostic information while running. Useful for debugging errors on large data sets.
#' 
kronos_predict = function(fit, period, time, factors, verbose = verbose){
  new_data <-  fit$xlevels
  
  new_data$zt     = seq(0, period, 0.25) 
  
  new_data = expand.grid(new_data)
  
  new_data$zt_cos = cos(new_data$zt * 2 * pi / period)
  new_data$zt_sin = sin(new_data$zt * 2 * pi / period)
  
  names(new_data)[names(new_data) == "zt"]     <- time
  names(new_data)[names(new_data) == "zt_cos"] <- paste0(time, "_cos")
  names(new_data)[names(new_data) == "zt_sin"] <- paste0(time, "_sin")
  new_data = cbind(new_data, y_hat = predict(fit, newdata = new_data))
  
  if(is.null(factors)){new_data$unique_group <- TRUE}
  
  if(!is.null(factors) & length(factors) == 1){
    new_data$unique_group = new_data[,{{factors}}]
  }
  
  if(!is.null(factors) & length(factors) > 1){
    new_data$unique_group = do.call(what = "paste", c(new_data[,{{factors}}], sep = "_"))
  }
  
  return(new_data)
}



#' Fit cosinor model
#' @description Fit cosinor model for one aspect of data. Called by main 'kronos' function. 
#' @param data input data
#' @param group A character string. Signifies which group will be assessed. 
#' @param time A string. Should be the column name containing the time values.  
#' @param period A numeric. The length of a period, in the same format as the \code{time} parameter.  
#' @param verbose A boolean. Toggles whether to print diagnostic information while running. Useful for debugging errors on large data sets.
#' 
fit_groupwise_model <- function(data, group, time, period, verbose){
  
  data = data[data[,"unique_group"] == group,]
  
  form = as.formula(paste0(colnames(data)[1], " ~ (", time, "_cos + ", time, "_sin)" ))
  
  if(verbose){print(paste0("Using the following model: ", c(form) ))}
  
  fit = lm(form, data = data)
  
  fstat = summary(fit)$fstatistic
  p.val = pf(fstat[1], fstat[2], fstat[3], lower.tail = FALSE)
  r.sq  = summary(fit)$r.squared
  
  vals = expand.grid(groups = group, zt = seq(0, period, 0.25))
  colnames(vals)[1] <- group
  
  vals$zt_cos = cos(vals$zt * 2 * pi / period)
  vals$zt_sin = sin(vals$zt * 2 * pi / period)
  
  
  vals$base     = coef(fit)[1]
  vals$cos_part = coef(fit)[paste0(time, "_cos")] * vals$zt_cos
  vals$sin_part = coef(fit)[paste0(time, "_sin")] * vals$zt_sin
  
  #sum them all up
  vals$pred_value = vals$base + vals$cos_part + vals$sin_part
  
  
  avg       = mean(vals$pred_value)
  #highpoint = which(vals$pred_value == max(vals$pred_value))
  sin_beta  = coef(fit)[paste0(time, "_sin")]
  cos_beta  = coef(fit)[paste0(time, "_cos")]
  acrophase = (atan2(sin_beta,cos_beta) / 2 / pi * period + period) %% period
  amplitude = sqrt(sin_beta^2 + cos_beta^2)
  
  out_df = data.frame(
    group     = group,
    p.val     = p.val,
    r.sq      = r.sq,
    avg       = avg,
    acro      = acrophase,
    amplitude = amplitude)
  
  colnames(out_df)[1] <- "unique_group"
  
  return(out_df)
}

#' Fit pairwise cosinor models as some sort of TukeyHSD.
#' @description Fit cosinor model for subset of data. Called by main 'kronos' function. 
#' @param data input data
#' @param formula A formula. Use the \code{time} function to designate which variable represents time. 
#' @param time A string. Should be the column name containing the time values.  
#' @param verbose A boolean. Toggles whether to print diagnostic information while running. Useful for debugging errors on large data sets.
#'
pairwise_cosinor_model <- function(data, formula, time, verbose){
  stopifnot("You need more that two groups in order to sensibly do pairwise comparisons. " = length(unique(data[,"unique_group"])) > 1)
  combos <- combn(c(unique(data[,"unique_group"])), m = 2)
  
  pairwise_models = list()
  
  for(c in 1:ncol(combos)){
    group_pair = combos[,c]
    x_pair     = data[data[,"unique_group"] %in% group_pair,]
    formula_pair = update.formula(formula, ~ unique_group)
    formula_pair = build_kronos_formula(formula = formula_pair, time = time)
    pairwise_models[[c]] <- anova(fit_cosinor_model(formula = formula_pair, data = x_pair, time = time, verbose = verbose, for_pw = TRUE))
    names(pairwise_models)[c] <- paste(group_pair, collapse = " vs ")
  }
  return(pairwise_models)
}

#' Update 'kronos' formula in light of sine and cosine components
#' @description Update 'kronos' formula.
#' @param formula A formula. Use the \code{time} function to designate which variable represents time. 
#' @param time A string. Should be the column name containing the time values.  
#' @param verbose A boolean. Toggles whether to print diagnostic information while running. Useful for debugging errors on large data sets.
#'
build_kronos_formula <- function(formula, time, verbose){
  #Trim off the time component
  temp_form = gsub("time\\(.*\\)", Reduce(paste0, deparse(formula)), replacement = "")
  temp_form = gsub("[ ]*\\*[ ]*$|[ ]*\\+[ ]*$", replacement = "", temp_form)
  
  #trim off trailing spaces
  temp_form = gsub("1$",    replacement = "", temp_form)
  temp_form = gsub("[ ]*$", replacement = "", temp_form)
  
  #if there were no other components to the formula:
  if(grepl("\\~$", x = temp_form)){
    return(formula(paste0(temp_form, " ", time, "_cos + ", time, "_sin")))
  }else{
  return(formula(paste0(temp_form, " * (", time, "_cos + ", time, "_sin)")))
    }
}

#' Extract p-value from full fit
#' @description Compute p-values from full fit. 
#' @param fit A lm model fit.  
#' @param time A string. Should be the column name containing the time values.  
#'
kronos_anova <- function(fit, time){
  anova.fit    <- anova(fit)
  time_int     <- grep(row.names(anova.fit), pattern = paste0(":", time, "_sin", "|", ":", time, "_cos"))
  pvals        <- anova.fit$`Pr(>F)`[time_int]
  pvals        <- pmin(pmin(pvals[1:length(pvals)%%2 ==0], pvals[1:length(pvals)%%2 ==1])*2, 1)
  names(pvals) <- unique(gsub(row.names(anova.fit), pattern = paste0(":", time, "_sin", "|", ":", time, "_cos"), replacement =  "")[time_int])
  return(pvals)
}


#' Fit a cosinor model and extract relevant parameters on a feature table.
#' @description This wrapper applies kronos(), the main workhorse function in the 'kronos' package. It manages the individual functionalities of 'kronos', including rhythmicity analysis and differential rhythmicity.
#' @param x Input data. A table with rows being features and columns being samples
#' @param formula A formula. Use the \code{time} function to designate which variable represents time. Leave the left-hand side of the formula empty as it will be sequentially replaced by every feature in the table.
#' @param metadata A metadata table, with rows being samples and columns being metadata entries
#' @param time A string. Should be the column name containing the time values.  
#' @param period A numeric. The length of a period, in the same format as the \code{time} parameter.  
#' @param pairwise A boolean. Toggles whether to perform pairwise ANOVAs as a TukeyHSD-like post-hoc. 
#' @param verbose A boolean. Toggles whether to print diagnostic information while running. Useful for debugging errors on large datasets.
#' @return A list of \code{kronosOut} S4 objects containing coefficients and all operations.
#' @importFrom methods new
#' @importFrom stats anova as.formula coef lm pf predict terms update.formula get_all_vars
#' @importFrom utils combn
#' @examples
#' #Load prepared data stored in 'kronos' library
#' data("kronos_demo")
#' output <- kronos(formula = Variable_1 ~ time(Timepoint), 
#' data = onevariable, period = 24, verbose = TRUE, pairwise = FALSE)
#' 
#' #Extracting data from the output object:
#' getKronos_fit(output)
#' getKronos_trace(output)
#' getKronos_groupwise(output)
#' 
#' #Plotting:
#' gg_kronos_circle(output)
#' gg_kronos_sinusoid(output)
#' 
#' #For high-dimensional data, use fw_kronos:
#' out_list = fw_kronos(x = bigdata[1:50,], formula = ~ Group + time(Timepoint), 
#' metadata = bigmeta, period = 24, verbose = FALSE, pairwise = TRUE) 
#'
#' #Extracting data from the output object: 
#' kronosListToTable(out_list)
#' 
#' 
#' #Plotting:
#' gg_kronos_acrogram(out_list)
#' @export
fw_kronos <- function(x, formula, metadata, time = NULL, period = 24, verbose = FALSE, pairwise = FALSE){
  formula = update.formula(x_feature ~ ., formula)
  apply(X = x, MARGIN = 1, FUN = function(y){
    kronos(formula = formula, 
           data = cbind("x_feature" = y, metadata),
           time = time, period = period,  
           verbose = verbose, 
           pairwise = pairwise)})
  
}
  